typing/typedecl.ml

```diff
@@ -91,12 +93,37 @@ let update_type temp_env env id loc =
       with Ctype.Unify trace ->
         raise (Error(loc, Type_clash (env, trace)))
 
-(* Determine if a type is (an abbreviation for) the type "float" *)
 (* We use the Ctype.expand_head_opt version of expand_head to get access
    to the manifest type of private abbreviations. *)
+let rec get_unboxed_type_representation env ty fuel =
+  if fuel < 0 then None else
+  let ty = Ctype.repr (Ctype.expand_head_opt env ty) in
+  match ty.desc with
+  | Tconstr (p, args, _) ->
+    let tydecl = Env.find_type p env in
+    if Builtin_attributes.has_unboxed tydecl.type_attributes then begin
+      match tydecl.type_kind with
+      | Type_record ([{ld_type = ty2; _}], _)
+      | Type_variant [{cd_args = Cstr_tuple [ty2]; _}]
+      | Type_variant [{cd_args = Cstr_record [{ld_type = ty2; _}]; _}]
+      -> get_unboxed_type_representation env
+           (Ctype.apply env tydecl.type_params ty2 args) (fuel - 1)
+      | Type_abstract -> None
+          (* This case can occur when checking a recursive unboxed type
+             declaration. *)
+      | _ -> assert false (* only the above can be unboxed *)
+    end else
+      Some ty
+  | _ -> Some ty
+
+let get_unboxed_type_representation env ty =
+  get_unboxed_type_representation env ty 100000
```

Comment:
Sorry to come back to that, but it can quickly take several seconds to type-check code such as:

````ocaml
type t = A of t [@@unboxed]
let f (x : t) = [| x |], [| x |], [| x |], [| x |], [| x |]
````

What about lowering significantly the "fuel" (e.g. to 100) and, to be on the safe side, reject declarations that "exhaust" the fuel?   Rejecting the declaration above would actually be better since no value can be built; and nobody will create chains of more than 100 nested "unboxed" types, right?